File: AleniaI_ntranet/src/main/java/com/iberia/intranet/service/LeaveService.java
package com.iberia.intranet.service;

import com.iberia.intranet.dto.LeaveBalanceDto;
import com.iberia.intranet.dto.LeaveRequestCreateDto;
import com.iberia.intranet.dto.LeaveRequestDto;
import com.iberia.intranet.dto.LeaveTypeDto;
import com.iberia.intranet.entity.LeaveStatus;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;
import java.util.UUID;

public interface LeaveService {

    // Leave Types
    List<LeaveTypeDto> getAllLeaveTypes();

    // Balances
    List<LeaveBalanceDto> getMyBalances();

    List<LeaveBalanceDto> getUserBalances(UUID userId);

    void initializeBalances(UUID userId, int year);

    // Requests
    LeaveRequestDto createRequest(LeaveRequestCreateDto createDto);

    LeaveRequestDto getRequest(UUID requestId);

    Page<LeaveRequestDto> getMyRequests(Pageable pageable);

    Page<LeaveRequestDto> getTeamRequests(Pageable pageable, LeaveStatus status);

    // Workflow
    LeaveRequestDto approveRequest(UUID requestId);

    LeaveRequestDto rejectRequest(UUID requestId, String reason);

    LeaveRequestDto cancelRequest(UUID requestId);

    // Calendar
    List<LeaveRequestDto> getMyTeamCalendar(int month, int year);
}

File: AleniaIntranetFront/src/app/core/services/leave.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpParams } from '@angular/common/http';
import { Observable } from 'rxjs';
import { LeaveBalance, LeaveRequest, LeaveRequestCreate, LeaveType, PageResponse, LeaveStatus } from '../models/leave.model';

@Injectable({
    providedIn: 'root'
})
export class LeaveService {
    private apiUrl = '/api/leaves';

    constructor(private http: HttpClient) { }

    getLeaveTypes(): Observable<LeaveType[]> {
        return this.http.get<LeaveType[]>(`${this.apiUrl}/types`);
    }

    getMyBalances(): Observable<LeaveBalance[]> {
        return this.http.get<LeaveBalance[]>(`${this.apiUrl}/my-balances`);
    }

    createRequest(request: LeaveRequestCreate): Observable<LeaveRequest> {
        return this.http.post<LeaveRequest>(`${this.apiUrl}/requests`, request);
    }

    getMyRequests(page: number = 0, size: number = 10): Observable<PageResponse<LeaveRequest>> {
        const params = new HttpParams()
            .set('page', page.toString())
            .set('size', size.toString())
            .set('sort', 'createdAt,desc');
        return this.http.get<PageResponse<LeaveRequest>>(`${this.apiUrl}/requests`, { params });
    }

    getRequest(id: string): Observable<LeaveRequest> {
        return this.http.get<LeaveRequest>(`${this.apiUrl}/requests/${id}`);
    }

    cancelRequest(id: string): Observable<LeaveRequest> {
        return this.http.put<LeaveRequest>(`${this.apiUrl}/requests/${id}/cancel`, {});
    }

    // Manager / Admin Endpoints

    getTeamRequests(page: number = 0, size: number = 10, status?: LeaveStatus): Observable<PageResponse<LeaveRequest>> {
        let params = new HttpParams()
            .set('page', page.toString())
            .set('size', size.toString())
            .set('sort', 'createdAt,desc');

        if (status) {
            params = params.set('status', status);
        }

        return this.http.get<PageResponse<LeaveRequest>>(`${this.apiUrl}/team/requests`, { params });
    }

    approveRequest(id: string): Observable<LeaveRequest> {
        return this.http.put<LeaveRequest>(`${this.apiUrl}/requests/${id}/approve`, {});
    }

    rejectRequest(id: string, reason: string): Observable<LeaveRequest> {
        return this.http.put<LeaveRequest>(`${this.apiUrl}/requests/${id}/reject`, { reason });
    }

    getUserBalances(userId: string): Observable<LeaveBalance[]> {
        return this.http.get<LeaveBalance[]>(`${this.apiUrl}/admin/users/${userId}/balances`);
    }
}
